File: vscode-extension/src/core/AgentCommunicationBus.ts
/**
 * AgentCommunicationBus - Inter-agent communication system
 * Enables agents to collaborate, share information, and coordinate actions
 */

import { EventEmitter } from 'events';

export interface AgentMessage {
    id: string;
    from: string;
    to: string | string[] | 'broadcast';
    type: MessageType;
    content: any;
    metadata: MessageMetadata;
    timestamp: number;
    replyTo?: string;
}

export enum MessageType {
    REQUEST = 'request',
    RESPONSE = 'response',
    NOTIFICATION = 'notification',
    QUERY = 'query',
    BROADCAST = 'broadcast',
    COLLABORATION_REQUEST = 'collaboration_request',
    COLLABORATION_RESPONSE = 'collaboration_response',
    TASK_DELEGATION = 'task_delegation',
    STATUS_UPDATE = 'status_update',
    ERROR = 'error',
    HELP_REQUEST = 'help_request',
    KNOWLEDGE_SHARE = 'knowledge_share',
    VALIDATION_REQUEST = 'validation_request',
    CONFLICT = 'conflict'
}

export interface MessageMetadata {
    priority: 'low' | 'normal' | 'high' | 'critical';
    requiresResponse: boolean;
    timeout?: number;
    conversationId?: string;
    workflowId?: string;
    retryCount?: number;
    confidence?: number;
    reasoning?: string;
}

export interface CollaborationSession {
    id: string;
    task: any;
    participants: string[];
    leader?: string;
    status: 'pending' | 'active' | 'completed' | 'failed';
    sharedContext: Map<string, any>;
    messages: AgentMessage[];
    results: Map<string, any>;
    startTime: number;
    endTime?: number;
}

export interface MessageHandler {
    agentId: string;
    messageTypes: MessageType[];
    handler: (message: AgentMessage) => Promise<any>;
    filter?: (message: AgentMessage) => boolean;
}

export interface CommunicationStats {
    totalMessages: number;
    messagesByType: Map<MessageType, number>;
    messagesByAgent: Map<string, number>;
    averageResponseTime: number;
    activeSessions: number;
    failedMessages: number;
}

export class AgentCommunicationBus {
    private static instance: AgentCommunicationBus;
    private eventBus: EventEmitter;
    private handlers: Map<string, MessageHandler[]> = new Map();
    private messageQueue: AgentMessage[] = [];
    private processingQueue: boolean = false;
    private collaborationSessions: Map<string, CollaborationSession> = new Map();
    private messageHistory: AgentMessage[] = [];
    private responseCallbacks: Map<string, (response: any) => void> = new Map();
    private stats: CommunicationStats;

    private constructor() {
        this.eventBus = new EventEmitter();
        this.eventBus.setMaxListeners(50);
        this.stats = this.initializeStats();
        this.startQueueProcessor();
    }

    public static getInstance(): AgentCommunicationBus {
        if (!AgentCommunicationBus.instance) {
            AgentCommunicationBus.instance = new AgentCommunicationBus();
        }
        return AgentCommunicationBus.instance;
    }

    /**
     * Register an agent to receive messages
     */
    public register(handler: MessageHandler): void {
        if (!this.handlers.has(handler.agentId)) {
            this.handlers.set(handler.agentId, []);
        }

        this.handlers.get(handler.agentId)!.push(handler);
        this.eventBus.emit('agent-registered', handler.agentId);
    }

    /**
     * Unregister an agent
     */
    public unregister(agentId: string): void {
        this.handlers.delete(agentId);
        this.eventBus.emit('agent-unregistered', agentId);
    }

    /**
     * Send a message to one or more agents
     */
    public async send(message: Omit<AgentMessage, 'id' | 'timestamp'>): Promise<string> {
        const fullMessage: AgentMessage = {
            ...message,
            id: this.generateMessageId(),
            timestamp: Date.now()
        };

        // Add to history
        this.messageHistory.push(fullMessage);
        this.stats.totalMessages++;
        this.updateStats(fullMessage);

        // Add to queue
        this.messageQueue.push(fullMessage);

        // Emit event
        this.eventBus.emit('message-sent', fullMessage);

        // Process queue if not already processing
        if (!this.processingQueue) {
            this.processQueue();
        }

        return fullMessage.id;
    }

    /**
     * Send a message and wait for response
     */
    public async request(
        message: Omit<AgentMessage, 'id' | 'timestamp'>,
        timeout: number = 30000
    ): Promise<any> {
        const messageId = await this.send({
            ...message,
            metadata: {
                ...(message.metadata || {}),
                requiresResponse: true,
                timeout
            }
        });

        return new Promise((resolve, reject) => {
            const timer = setTimeout(() => {
                this.responseCallbacks.delete(messageId);
                reject(new Error(`Request timeout for message ${messageId}`));
            }, timeout);

            this.responseCallbacks.set(messageId, (response) => {
                clearTimeout(timer);
                this.responseCallbacks.delete(messageId);
                resolve(response);
            });
        });
    }

    /**
     * Broadcast a message to all agents
     */
    public async broadcast(
        from: string,
        type: MessageType,
        content: any,
        metadata?: Partial<MessageMetadata>
    ): Promise<void> {
        await this.send({
            from,
            to: 'broadcast',
            type: MessageType.BROADCAST,
            content,
            metadata: {
                priority: 'normal',
                requiresResponse: false,
                ...metadata
            }
        });
    }

    /**
     * Start a collaboration session between agents
     */
    public async startCollaboration(
        task: any,
        participants: string[],
        leader?: string
    ): Promise<CollaborationSession> {
        const session: CollaborationSession = {
            id: this.generateSessionId(),
            task,
            participants,
            leader: leader || participants[0],
            status: 'pending',
            sharedContext: new Map(),
            messages: [],
            results: new Map(),
            startTime: Date.now()
        };

        this.collaborationSessions.set(session.id, session);

        // Notify all participants
        await Promise.all(participants.map(agentId =>
            this.send({
                from: 'system',
                to: agentId,
                type: MessageType.COLLABORATION_REQUEST,
                content: {
                    sessionId: session.id,
                    task,
                    participants,
                    leader: session.leader
                },
                metadata: {
                    priority: 'high',
                    requiresResponse: true,
                    conversationId: session.id
                }
            })
        ));

        session.status = 'active';
        this.eventBus.emit('collaboration-started', session);

        return session;
    }

    /**
     * Send a message within a collaboration session
     */
    public async collaborationMessage(
        sessionId: string,
        from: string,
        content: any,
        type: MessageType = MessageType.NOTIFICATION
    ): Promise<void> {
        const session = this.collaborationSessions.get(sessionId);
        if (!session) {
            throw new Error(`Collaboration session ${sessionId} not found`);
        }

        // Send to all participants except sender
        const recipients = session.participants.filter(p => p !== from);

        const message = {
            from,
            to: recipients,
            type,
            content,
            metadata: {
                priority: 'normal' as const,
                requiresResponse: false,
                conversationId: sessionId
            }
        };

        await this.send(message);

        // Add to session history
        session.messages.push({
            ...message,
            id: this.generateMessageId(),
            timestamp: Date.now()
        });
    }

    /**
     * Update shared context in collaboration session
     */
    public updateCollaborationContext(
        sessionId: string,
        agentId: string,
        key: string,
        value: any
    ): void {
        const session = this.collaborationSessions.get(sessionId);
        if (!session) return;

        session.sharedContext.set(key, value);

        // Notify other participants
        this.collaborationMessage(
            sessionId,
            agentId,
            { key, value },
            MessageType.STATUS_UPDATE
        );
    }

    /**
     * Complete a collaboration session
     */
    public completeCollaboration(sessionId: string, results: Map<string, any>): void {
        const session = this.collaborationSessions.get(sessionId);
        if (!session) return;

        session.status = 'completed';
        session.results = results;
        session.endTime = Date.now();

        this.eventBus.emit('collaboration-completed', session);

        // Clean up after delay
        setTimeout(() => {
            this.collaborationSessions.delete(sessionId);
        }, 60000);
    }

    /**
     * Request help from other agents
     */
    public async requestHelp(
        from: string,
        problem: any,
        preferredAgents?: string[]
    ): Promise<any> {
        const message = {
            from,
            to: preferredAgents || 'broadcast',
            type: MessageType.HELP_REQUEST,
            content: problem,
            metadata: {
                priority: 'high' as const,
                requiresResponse: true,
                timeout: 10000
            }
        };

        const responses: any[] = [];

        if (preferredAgents) {
            // Request from specific agents
            for (const agentId of preferredAgents) {
                try {
                    const response = await this.request(
                        { ...message, to: agentId, metadata: message.metadata },
                        10000
                    );
                    if (response) responses.push(response);
                } catch (error) {
                    console.warn(`No response from ${agentId}:`, error);
                }
            }
        } else {
            // Broadcast and collect responses
            await this.broadcast(from, MessageType.HELP_REQUEST, problem, {
                priority: 'high',
                requiresResponse: true
            });

            // Wait for responses
            await new Promise(resolve => setTimeout(resolve, 5000));

            // Collect responses from history
            const requestTime = Date.now();
            responses.push(...this.messageHistory
                .filter(msg =>
                    msg.type === MessageType.RESPONSE &&
                    msg.timestamp > requestTime - 5000 &&
                    msg.replyTo === message.from
                )
                .map(msg => msg.content)
            );
        }

        return responses;
    }

    /**
     * Share knowledge between agents
     */
    public async shareKnowledge(
        from: string,
        knowledge: any,
        relevantAgents?: string[]
    ): Promise<void> {
        await this.send({
            from,
            to: relevantAgents || 'broadcast',
            type: MessageType.KNOWLEDGE_SHARE,
            content: knowledge,
            metadata: {
                priority: 'low',
                requiresResponse: false
            }
        });
    }

    /**
     * Request validation from another agent
     */
    public async requestValidation(
        from: string,
        validator: string,
        content: any
    ): Promise<any> {
        return this.request({
            from,
            to: validator,
            type: MessageType.VALIDATION_REQUEST,
            content,
            metadata: {
                priority: 'normal',
                requiresResponse: true
            }
        }, 15000);
    }

    /**
     * Report a conflict requiring arbitration
     */
    public async reportConflict(
        reportingAgent: string,
        conflictingAgents: string[],
        issue: any
    ): Promise<void> {
        // Send to OpusArbitrator
        await this.send({
            from: reportingAgent,
            to: 'OpusArbitrator',
            type: MessageType.CONFLICT,
            content: {
                conflictingAgents,
                issue,
                reportedBy: reportingAgent
            },
            metadata: {
                priority: 'critical',
                requiresResponse: true
            }
        });
    }

    /**
     * Process message queue
     */
    private async processQueue(): Promise<void> {
        if (this.processingQueue || this.messageQueue.length === 0) return;

        this.processingQueue = true;

        while (this.messageQueue.length > 0) {
            const message = this.messageQueue.shift()!;

            try {
                await this.deliverMessage(message);
            } catch (error) {
                console.error(`Error delivering message ${message.id}:`, error);
                this.stats.failedMessages++;

                // Retry logic
                if (message.metadata.retryCount === undefined) {
                    message.metadata.retryCount = 0;
                }

                if (message.metadata.retryCount < 3) {
                    message.metadata.retryCount++;
                    this.messageQueue.push(message);
                } else {
                    this.eventBus.emit('message-failed', { message, error });
                }
            }
        }

        this.processingQueue = false;
    }

    /**
     * Deliver a message to recipients
     */
    private async deliverMessage(message: AgentMessage): Promise<void> {
        const recipients = this.determineRecipients(message);

        for (const recipientId of recipients) {
            const handlers = this.handlers.get(recipientId) || [];

            for (const handler of handlers) {
                // Check if handler accepts this message type
                if (!handler.messageTypes.includes(message.type)) continue;

                // Apply filter if present
                if (handler.filter && !handler.filter(message)) continue;

                try {
                    const response = await handler.handler(message);

                    // Handle response if required
                    if (message.metadata.requiresResponse && response !== undefined) {
                        // Send response back
                        await this.send({
                            from: recipientId,
                            to: message.from,
                            type: MessageType.RESPONSE,
                            content: response,
                            metadata: {
                                priority: 'normal',
                                requiresResponse: false,
                                conversationId: message.metadata.conversationId
                            },
                            replyTo: message.id
                        });

                        // Trigger callback if waiting
                        const callback = this.responseCallbacks.get(message.id);
                        if (callback) {
                            callback(response);
                        }
                    }

                    this.eventBus.emit('message-delivered', { message, recipientId });
                } catch (error) {
                    console.error(`Handler error for ${recipientId}:`, error);
                    this.eventBus.emit('handler-error', { message, recipientId, error });
                }
            }
        }
    }

    /**
     * Determine message recipients
     */
    private determineRecipients(message: AgentMessage): string[] {
        if (message.to === 'broadcast') {
            return Array.from(this.handlers.keys());
        }

        if (Array.isArray(message.to)) {
            return message.to;
        }

        return [message.to];
    }

    /**
     * Start queue processor
     */
    private startQueueProcessor(): void {
        setInterval(() => {
            if (!this.processingQueue && this.messageQueue.length > 0) {
                this.processQueue();
            }
        }, 100);
    }

    /**
     * Initialize statistics
     */
    private initializeStats(): CommunicationStats {
        return {
            totalMessages: 0,
            messagesByType: new Map(),
            messagesByAgent: new Map(),
            averageResponseTime: 0,
            activeSessions: 0,
            failedMessages: 0
        };
    }

    /**
     * Update statistics
     */
    private updateStats(message: AgentMessage): void {
        // Update message type count
        const typeCount = this.stats.messagesByType.get(message.type) || 0;
        this.stats.messagesByType.set(message.type, typeCount + 1);

        // Update agent message count
        const agentCount = this.stats.messagesByAgent.get(message.from) || 0;
        this.stats.messagesByAgent.set(message.from, agentCount + 1);

        // Update active sessions
        this.stats.activeSessions = this.collaborationSessions.size;
    }

    /**
     * Get communication statistics
     */
    public getStats(): CommunicationStats {
        // Calculate average response time
        let totalResponseTime = 0;
        let responseCount = 0;

        this.messageHistory.forEach(msg => {
            if (msg.type === MessageType.RESPONSE && msg.replyTo) {
                const originalMsg = this.messageHistory.find(m => m.id === msg.replyTo);
                if (originalMsg) {
                    totalResponseTime += msg.timestamp - originalMsg.timestamp;
                    responseCount++;
                }
            }
        });

        this.stats.averageResponseTime = responseCount > 0
            ? totalResponseTime / responseCount
            : 0;

        return { ...this.stats };
    }

    /**
     * Get message history
     */
    public getMessageHistory(filter?: {
        from?: string;
        to?: string;
        type?: MessageType;
        conversationId?: string;
        startTime?: number;
        endTime?: number;
    }): AgentMessage[] {
        let history = [...this.messageHistory];

        if (filter) {
            if (filter.from) {
                history = history.filter(msg => msg.from === filter.from);
            }
            if (filter.to) {
                history = history.filter(msg =>
                    msg.to === filter.to ||
                    (Array.isArray(msg.to) && msg.to.includes(filter.to!))
                );
            }
            if (filter.type) {
                history = history.filter(msg => msg.type === filter.type);
            }
            if (filter.conversationId) {
                history = history.filter(msg =>
                    msg.metadata.conversationId === filter.conversationId
                );
            }
            if (filter.startTime) {
                history = history.filter(msg => msg.timestamp >= filter.startTime!);
            }
            if (filter.endTime) {
                history = history.filter(msg => msg.timestamp <= filter.endTime!);
            }
        }

        return history;
    }

    /**
     * Get active collaboration sessions
     */
    public getActiveSessions(): CollaborationSession[] {
        return Array.from(this.collaborationSessions.values())
            .filter(session => session.status === 'active');
    }

    /**
     * Generate unique message ID
     */
    private generateMessageId(): string {
        return `msg_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
    }

    /**
     * Generate unique session ID
     */
    private generateSessionId(): string {
        return `ses_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
    }

    /**
     * Clear all data (for testing/reset)
     */
    public clear(): void {
        this.messageQueue = [];
        this.messageHistory = [];
        this.collaborationSessions.clear();
        this.responseCallbacks.clear();
        this.stats = this.initializeStats();
    }
}

// Export singleton instance getter
export function getCommunicationBus(): AgentCommunicationBus {
    return AgentCommunicationBus.getInstance();
}

File: backend/core/agent_communication_bus.py
"""
AgentCommunicationBus - Event-driven communication system for inter-agent messaging
Enables agents to collaborate, share knowledge, and request help
"""

import asyncio
import time
import uuid
from typing import Dict, List, Optional, Any, Callable, Set, Tuple
from dataclasses import dataclass, field
from enum import Enum
from collections import defaultdict
import logging

logger = logging.getLogger(__name__)

class MessageType(Enum):
    """Types of inter-agent messages"""
    REQUEST = "request"              # Request for collaboration
    RESPONSE = "response"            # Response to request
    BROADCAST = "broadcast"          # Broadcast to all agents
    HELP_REQUEST = "help_request"    # Request for help
    KNOWLEDGE_SHARE = "knowledge_share"  # Share knowledge/patterns
    STATUS_UPDATE = "status_update"  # Status update
    CONFLICT = "conflict"            # Conflict notification
    VOTE = "vote"                    # Voting on decision

@dataclass
class AgentMessage:
    """Message between agents"""
    id: str
    from_agent: str
    to_agent: str  # 'all' for broadcast
    message_type: MessageType
    content: Any
    timestamp: float = field(default_factory=time.time)
    correlation_id: Optional[str] = None
    priority: int = 0  # Higher is more important
    expires_at: Optional[float] = None
    metadata: Dict[str, Any] = field(default_factory=dict)

@dataclass
class MessageHandler:
    """Handler for messages"""
    agent_id: str
    callback: Callable
    filter: Optional[Callable] = None
    message_types: Optional[List[MessageType]] = None

@dataclass
class CollaborationSession:
    """Active collaboration between agents"""
    id: str
    initiator: str
    participants: List[str]
    topic: str
    started_at: float
    messages: List[AgentMessage]
    status: str  # 'active', 'completed', 'failed'

class AgentCommunicationBus:
    """
    Central communication bus for inter-agent messaging
    """

    _instance = None

    def __new__(cls):
        """Singleton pattern"""
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance._initialized = False
        return cls._instance

    def __init__(self):
        """Initialize communication bus"""
        if self._initialized:
            return

        # Message routing
        self.handlers: Dict[str, List[MessageHandler]] = defaultdict(list)
        self.topics: Dict[str, Set[str]] = defaultdict(set)  # topic -> subscribers

        # Message storage
        self.message_queue: asyncio.Queue = asyncio.Queue()
        self.message_history: List[AgentMessage] = []
        self.pending_responses: Dict[str, asyncio.Future] = {}

        # Collaboration sessions
        self.collaboration_sessions: Dict[str, CollaborationSession] = {}

        # Statistics
        self.total_messages = 0
        self.messages_by_type = defaultdict(int)
        self.messages_by_agent = defaultdict(int)

        # Background task for message processing
        self.processing_task = None

        self._initialized = True
        logger.info("AgentCommunicationBus initialized")

    async def start(self):
        """Start the message processing loop"""
        if not self.processing_task:
            self.processing_task = asyncio.create_task(self._process_messages())
            logger.info("Communication bus started")

    async def stop(self):
        """Stop the message processing loop"""
        if self.processing_task:
            self.processing_task.cancel()
            try:
                await self.processing_task
            except asyncio.CancelledError:
                pass
            self.processing_task = None
            logger.info("Communication bus stopped")

    async def register_agent(
        self,
        agent_id: str,
        handler: Callable,
        message_types: Optional[List[MessageType]] = None,
        topics: Optional[List[str]] = None
    ):
        """Register an agent with the communication bus"""
        # Register message handler
        message_handler = MessageHandler(
            agent_id=agent_id,
            callback=handler,
            message_types=message_types
        )

        self.handlers[agent_id].append(message_handler)

        # Subscribe to topics
        if topics:
            for topic in topics:
                self.topics[topic].add(agent_id)

        logger.info(f"Agent {agent_id} registered with communication bus")

    async def unregister_agent(self, agent_id: str):
        """Unregister an agent"""
        # Remove handlers
        if agent_id in self.handlers:
            del self.handlers[agent_id]

        # Remove from topics
        for topic_agents in self.topics.values():
            topic_agents.discard(agent_id)

        logger.info(f"Agent {agent_id} unregistered from communication bus")

    async def send_message(
        self,
        from_agent: str,
        to_agent: str,
        message_type: MessageType,
        content: Any,
        correlation_id: Optional[str] = None,
        priority: int = 0,
        timeout: Optional[float] = None
    ) -> Optional[Any]:
        """Send a message from one agent to another"""
        message_id = self._generate_id()

        message = AgentMessage(
            id=message_id,
            from_agent=from_agent,
            to_agent=to_agent,
            message_type=message_type,
            content=content,
            correlation_id=correlation_id or message_id,
            priority=priority,
            expires_at=time.time() + timeout if timeout else None
        )

        # Add to queue
        await self.message_queue.put(message)
        self.total_messages += 1
        self.messages_by_type[message_type] += 1
        self.messages_by_agent[from_agent] += 1

        # If expecting response, create future
        if message_type == MessageType.REQUEST:
            future = asyncio.Future()
            self.pending_responses[message.correlation_id] = future

            try:
                # Wait for response with timeout
                response = await asyncio.wait_for(
                    future,
                    timeout=timeout or 30.0
                )
                return response
            except asyncio.TimeoutError:
                del self.pending_responses[message.correlation_id]
                logger.warning(f"Request from {from_agent} to {to_agent} timed out")
                return None

        return message_id

    async def broadcast(
        self,
        from_agent: str,
        message_type: MessageType,
        content: Any,
        topic: Optional[str] = None
    ):
        """Broadcast a message to all agents or topic subscribers"""
        message = AgentMessage(
            id=self._generate_id(),
            from_agent=from_agent,
            to_agent="all",
            message_type=message_type,
            content=content,
            metadata={"topic": topic} if topic else {}
        )

        await self.message_queue.put(message)
        self.total_messages += 1
        self.messages_by_type[message_type] += 1
        self.messages_by_agent[from_agent] += 1

        logger.debug(f"Broadcast from {from_agent}: {message_type.value}")

    async def request_help(
        self,
        from_agent: str,
        task_description: str,
        required_capabilities: Optional[List[str]] = None
    ) -> List[Tuple[str, Any]]:
        """Request help from other agents"""
        help_id = self._generate_id()

        content = {
            "task": task_description,
            "required_capabilities": required_capabilities or [],
            "help_id": help_id
        }

        # Broadcast help request
        await self.broadcast(
            from_agent=from_agent,
            message_type=MessageType.HELP_REQUEST,
            content=content
        )

        # Wait for responses
        await asyncio.sleep(2.0)  # Give agents time to respond

        # Collect responses
        responses = []
        for message in self.message_history[-20:]:  # Check recent messages
            if (message.message_type == MessageType.RESPONSE and
                message.correlation_id == help_id):
                responses.append((message.from_agent, message.content))

        logger.info(f"Agent {from_agent} received {len(responses)} help responses")
        return responses

    async def start_collaboration(
        self,
        initiator: str,
        participants: List[str],
        topic: str
    ) -> str:
        """Start a collaboration session"""
        session_id = self._generate_id()

        session = CollaborationSession(
            id=session_id,
            initiator=initiator,
            participants=participants,
            topic=topic,
            started_at=time.time(),
            messages=[],
            status="active"
        )

        self.collaboration_sessions[session_id] = session

        # Notify participants
        for participant in participants:
            await self.send_message(
                from_agent=initiator,
                to_agent=participant,
                message_type=MessageType.REQUEST,
                content={
                    "action": "join_collaboration",
                    "session_id": session_id,
                    "topic": topic
                }
            )

        logger.info(f"Collaboration session {session_id} started: {topic}")
        return session_id

    async def end_collaboration(self, session_id: str, status: str = "completed"):
        """End a collaboration session"""
        if session_id in self.collaboration_sessions:
            session = self.collaboration_sessions[session_id]
            session.status = status

            # Notify participants
            for participant in session.participants:
                await self.send_message(
                    from_agent=session.initiator,
                    to_agent=participant,
                    message_type=MessageType.STATUS_UPDATE,
                    content={
                        "action": "collaboration_ended",
                        "session_id": session_id,
                        "status": status
                    }
                )

            logger.info(f"Collaboration session {session_id} ended: {status}")

    async def share_knowledge(
        self,
        from_agent: str,
        knowledge_type: str,
        knowledge_content: Any,
        recipients: Optional[List[str]] = None
    ):
        """Share knowledge with other agents"""
        content = {
            "knowledge_type": knowledge_type,
            "content": knowledge_content,
            "shared_by": from_agent,
            "timestamp": time.time()
        }

        if recipients:
            # Send to specific agents
            for recipient in recipients:
                await self.send_message(
                    from_agent=from_agent,
                    to_agent=recipient,
                    message_type=MessageType.KNOWLEDGE_SHARE,
                    content=content
                )
        else:
            # Broadcast to all
            await self.broadcast(
                from_agent=from_agent,
                message_type=MessageType.KNOWLEDGE_SHARE,
                content=content
            )

    async def vote_on_decision(
        self,
        topic: str,
        options: List[Any],
        voters: List[str],
        timeout: float = 10.0
    ) -> Dict[Any, int]:
        """Conduct a vote among agents"""
        vote_id = self._generate_id()

        # Send vote request to all voters
        for voter in voters:
            await self.send_message(
                from_agent="system",
                to_agent=voter,
                message_type=MessageType.VOTE,
                content={
                    "vote_id": vote_id,
                    "topic": topic,
                    "options": options
                },
                correlation_id=vote_id
            )

        # Wait for votes
        await asyncio.sleep(timeout)

        # Count votes
        votes = defaultdict(int)
        for message in self.message_history[-50:]:
            if (message.message_type == MessageType.RESPONSE and
                message.correlation_id == vote_id):
                vote = message.content.get("vote")
                if vote in options:
                    votes[vote] += 1

        logger.info(f"Vote on '{topic}' completed: {dict(votes)}")
        return dict(votes)

    async def _process_messages(self):
        """Background task to process messages"""
        while True:
            try:
                # Get message from queue
                message = await self.message_queue.get()

                # Check if expired
                if message.expires_at and time.time() > message.expires_at:
                    logger.debug(f"Message {message.id} expired")
                    continue

                # Store in history
                self.message_history.append(message)
                if len(self.message_history) > 1000:
                    self.message_history = self.message_history[-500:]

                # Route message
                await self._route_message(message)

                # Handle collaboration sessions
                if message.correlation_id in self.collaboration_sessions:
                    session = self.collaboration_sessions[message.correlation_id]
                    session.messages.append(message)

            except asyncio.CancelledError:
                break
            except Exception as e:
                logger.error(f"Error processing message: {e}")

    async def _route_message(self, message: AgentMessage):
        """Route message to appropriate handlers"""
        # Handle broadcast
        if message.to_agent == "all":
            # Send to all agents except sender
            for agent_id, handlers in self.handlers.items():
                if agent_id != message.from_agent:
                    await self._deliver_to_handlers(message, handlers)

        # Handle topic broadcast
        elif message.metadata.get("topic"):
            topic = message.metadata["topic"]
            if topic in self.topics:
                for agent_id in self.topics[topic]:
                    if agent_id != message.from_agent and agent_id in self.handlers:
                        await self._deliver_to_handlers(message, self.handlers[agent_id])

        # Handle direct message
        elif message.to_agent in self.handlers:
            await self._deliver_to_handlers(message, self.handlers[message.to_agent])

        # Handle response to pending request
        if message.message_type == MessageType.RESPONSE and message.correlation_id:
            if message.correlation_id in self.pending_responses:
                future = self.pending_responses[message.correlation_id]
                if not future.done():
                    future.set_result(message.content)
                del self.pending_responses[message.correlation_id]

    async def _deliver_to_handlers(
        self,
        message: AgentMessage,
        handlers: List[MessageHandler]
    ):
        """Deliver message to handlers"""
        for handler in handlers:
            # Check if handler wants this message type
            if handler.message_types and message.message_type not in handler.message_types:
                continue

            # Apply filter if present
            if handler.filter and not handler.filter(message):
                continue

            # Deliver message
            try:
                if asyncio.iscoroutinefunction(handler.callback):
                    await handler.callback(message)
                else:
                    handler.callback(message)
            except Exception as e:
                logger.error(f"Error in message handler: {e}")

    def get_stats(self) -> Dict[str, Any]:
        """Get communication statistics"""
        return {
            "total_messages": self.total_messages,
            "messages_by_type": dict(self.messages_by_type),
            "messages_by_agent": dict(self.messages_by_agent),
            "active_collaborations": len([
                s for s in self.collaboration_sessions.values()
                if s.status == "active"
            ]),
            "pending_responses": len(self.pending_responses),
            "registered_agents": len(self.handlers)
        }

    def _generate_id(self) -> str:
        """Generate unique ID"""
        return f"msg_{uuid.uuid4().hex[:8]}"

# Global instance
def get_communication_bus() -> AgentCommunicationBus:
    """Get singleton AgentCommunicationBus instance"""
    return AgentCommunicationBus()